### DISCUSSION TITLE
Invalid PID Sequence

### DISCUSSION THREAD

**Discussion by wocard2** [ataradov/usb-sniffer#17] [General]
Capturing low-speed USB works just fine, but for a high-speed device, I see only packets sent from the host to the device (no packets sent from the device to the host) and some ACK packets that look like this:

Frame 240: 1 byte on wire (8 bits), 1 byte captured (8 bits) on interface usb, id 0
    Section number: 1
    Interface id: 0 (usb)
    Encapsulation type: High-Speed USB 2.0 packets (217)
    Arrival Time: Jan  1, 1970 01:00:10.484624116 CET
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 10.484624116 seconds
    [Time delta from previous captured frame: 0.000009416 seconds]
    [Time delta from previous displayed frame: 0.000009416 seconds]
    [Time since reference or first frame: 10.484624116 seconds]
    Frame Number: 240
    Frame Length: 1 byte (8 bits)
    Capture Length: 1 byte (8 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: usbll]
USB Link Layer
    PID: ACK (0xd2)
        [Expert Info (Error/Malformed): Invalid PID Sequence]
            [Invalid PID Sequence]
            [Severity level: Error]
            [Group: Malformed]

Haven't tried a full-speed device yet.

What could be wrong with my build of your USB sniffer?

**Comment by ataradov** [comment]
What OS you are using? Did you set the correct speed in the settings?

Normally there is no difference for the sniffer which side sends the data. So, I'm inclined to think that there may be a signal integrity issue due to bad cables or connectors. Does the device under test work as expected when connected though the sniffer?

**Reply by wocard2** [reply]
The OS is Linux. I set the speed to high speed. The DUT (a flash drive) does work when connected through the sniffer.

**Reply by ataradov** [reply]
Capture a small, but complete log from plugging in the drive.
If only one side is visible, it must be some sort of signal integrity issue. I would definitely try with different cables. 
Another possible option is incorrect clock for the USB PHY, but it must be just a bit off for anything to be received at all.

**Comment by wocard2** [comment]
I see the chirp, then SOF, IN, and the ACK packet shown above (with "Invalid PID sequence"), followed by more IN, SOF, and ACK packets of the same type.

**Comment by ataradov** [comment]
I would try with different devices to collect some statistics on what works and what does not. It may be specific to that device.

**Comment by wocard2** [comment]
[hs-storage-bad2.pcapng.gz](https://github.com/ataradov/usb-sniffer/files/12306608/hs-storage-bad2.pcapng.gz)
I already tried different cables.

**Comment by ataradov** [comment]
I would say this is some sort of a clock issue. Only very short packets are received. Host would also send a lot of SETUP packets, and you can't see a lot of those.

What clock generator IC did you use? Also, what speed grade FPGA? I would check the frequency and stability of that clock and of the resulting 60 MHz output clock.

**Comment by wocard2** [comment]
Maybe a problem with the capacitors? As 1 µF capacitors that don't lose a lot of capacitance at 5 V in 0603 packages are hard to find, I used 50 V 0805 capacitors for the LDO, but I left all the other capacitors at 0603. For the 2.2 µF 0603 caps I had only 16 V specimen in stock, maybe I should replace these.

https://pdfserv.maximintegrated.com/en/an/[TUT5527.pdf](https://pdfserv.maximintegrated.com/en/an/TUT5527.pdf

**Comment by wocard2** [comment]
Speed grade 5 for the FPGA, the 26 MHz oscillator is a Yangxing Tech OT2EL4C4JI-111OLP-26M (sorry about that) with ±10 PPM. Haven't measured the 60 MHz clock yet, only checked if it's present. Apart from my 200 MHz scope I only have a 10 MHz frequency counter (with TCXO), I'd need a divider for that.

**Comment by ataradov** [comment]
I doubt it. I did not do anything special for the capacitors, just used nominal values.  There are a lot more capacitors than actually necessary on that board, partially to account for potential DC loss. 

This is either clocking issue, or PHY damage. PHY may have damage in the HS RX path only, for example.

I would just try with a lot more devices, including FS and see if there are any patterns show up.

**Reply by wocard2** [reply]
The LDO has only the minimum values for the caps nearby and some 0603 50(!) V caps from reputable manufacturers lose 40% of their capacitance at 5 V.

**Comment by ataradov** [comment]
Is 3.3 V power rail stable? If so, then capacitance it enough. You are really looking at it from the wrong end.

I've used that series of oscillators for 12 MHz and they worked fine, I have no reason to suspect they are bad.

Do more experiments with different devices.

**Comment by wocard2** [comment]
The T_USB_CLK frequency is 60.0006 MHz according to my scope. No jitter visible.

**Comment by ataradov** [comment]
Try with different devices, including FS. 

I would suspect some PHY damage. No idea why it might happen.

Replace the PHY and see if that helps.

**Comment by wocard2** [comment]
Perhaps the USB3343 got too hot during soldering.

**Comment by wocard2** [comment]
I tried a different storage device: this time I see SCSI packets!

**Comment by wocard2** [comment]
However, there still are ACK packages with "Invalid PID Sequence".

**Comment by wocard2** [comment]
While capturing, the 3.3 V rail has 3.3007 V (multimeter) with 12 mV peak-to-peak noise (scope), measured across the EEPROM, that was the only place I found where I can conveniently use a scope probe without resorting to using that ground wire aka antenna.

**Comment by ataradov** [comment]
Your old log also had a lot of DATAx packets and some responses from the device too. That log actually contains two enumerations, the last one for something called "MyUSB Drive". The enumeration process happens around frame 16000 - 17000 and it seems to go just fine.

That flash drive seems to be in general really slow to respond, but it responds and enumerates fine.

Invalid sequence happens because some DATAx packets are missing. This is "normal" until missing packets are resolved.

**Comment by wocard2** [comment]
HS flash drive A: no data from device
HS flash drive B: some data from device, still ACK errors
HS FX2-based logic analyzer: no data from device
FS STM32F042: works

**Comment by wocard2** [comment]
So I just didn't scroll far enough?

**Comment by ataradov** [comment]
Did you do the speed test for the sniffer? What were the results?

There is still something wrong. The data is captured correctly, but seemingly intermittently. And I don't understand how this is possible. I would expect to see a lot more broken packets. But there are none, all received packets are intact.

**Comment by wocard2** [comment]
The speed is a bit lowish:
Transfer rate: 40.37 MB/s
Transfer rate: 38.39 MB/s
Transfer rate: 39.37 MB/s
Transfer rate: 40.35 MB/s
Transfer rate: 41.33 MB/s
Transfer rate: 39.72 MB/s

**Comment by wocard2** [comment]
A USB 3.0 port is better:
Transfer rate: 43.44 MB/s
Transfer rate: 43.08 MB/s
Transfer rate: 43.17 MB/s
Transfer rate: 43.56 MB/s
Transfer rate: 43.28 MB/s
Transfer rate: 43.27 MB/s

**Comment by wocard2** [comment]
And here's my best USB 3 port:
Transfer rate: 48.84 MB/s
Transfer rate: 49.06 MB/s
Transfer rate: 49.11 MB/s
Transfer rate: 49.09 MB/s
Transfer rate: 49.10 MB/s

**Comment by wocard2** [comment]
With this (last) USB 3 port, HS flash drive A looks good!

**Comment by wocard2** [comment]
I thought messages were logged if the FX2 drops packets. Haven't seen any such message.

**Comment by ataradov** [comment]
What is your hardware setup? It looks like there is a hub somewhere because even in your initial log I see hub commands to reset the ports.

Ideally you want to use the root hub interface with no additional hubs, since that degrades performance quite a bit. But I'm still not sure why packets would go missing like this. Even if there is an overflow, there must be a message about that. Strange.

**Comment by wocard2** [comment]
[hs-storage-good2.pcapng.gz](https://github.com/ataradov/usb-sniffer/files/12307014/hs-storage-good2.pcapng.gz)
Here's flash drive A mounted and unmounted. Any packets missing?

**Comment by wocard2** [comment]
Speed test 1: USB 2.0 at root hub
Speed test 2: USB 3.0 probably with some internal hub
Speed test 3: USB 3.0 probably at root hub

**Comment by wocard2** [comment]
With USB, there are always hub commands as there's always a hub.

**Comment by wocard2** [comment]
![test1](https://github.com/ataradov/usb-sniffer/assets/52967668/e752b727-32bb-43b4-a3dc-b86e7321d965)
![test2](https://github.com/ataradov/usb-sniffer/assets/52967668/74bb9e6a-c068-4b73-a822-e288bd1fa9eb)
![test3](https://github.com/ataradov/usb-sniffer/assets/52967668/55071a53-bcfb-47e6-b1a7-a55c6a5eaef4)
Here are screenshots from usbview for the three speed tests. Interestingly, the fastest test has one hub more than the second-to-fastest test.

**Comment by ataradov** [comment]
This is exactly how it should look like (the log I mean). 

There are not always hub commands, even the sample HS log in that repo does not have them. This was captured from my PC with a sniffer attached to the root hub. Root hub does not have ports, so requests to reset a port are meaningless. 
This is some implementation detail of that host controller.

**Comment by wocard2** [comment]
(These are three images, "MyUSB Drive" is where the USB sniffer was connected for the respective speed test.)

**Comment by wocard2** [comment]
So, avoiding hubs isn't always the solution (see speed test 3 vs. speed test 2).

**Comment by ataradov** [comment]
What PC / motherboard is this? There is something in its architecture that does not make sense.

Using the fastest port is always a good idea, that's why there is a speed test command. But I don't have any explanation for the observed results.

This may be also some OS/drivers issue. Or those host controllers connected in different ways resulting in different performance.

**Comment by wocard2** [comment]
Interestingly, usbview claims that that fastest xHCI controller (test 3) runs at 480 Mb/s and the "middle" xHCI controller (test 2) runs at 5 Gb/s. I think usbview confused those two xHCI controllers. That way, the results make much more sense.

The motherboard is quite old, it uses the Z77 chipset. But I have some additional USB 3.0 host adapters installed, the USB port that turned out to be fastest may be connected to one of these.

**Comment by wocard2** [comment]
BTW, I'm using an MCP1755T-3302E/O instead of a MIC5504-3.3v. It's more expensive, but I already had some of them. Shouldn't make a difference and the measured 3.3 V rail looks good.

**Comment by wocard2** [comment]
Ah, I misread the indentation in the third image (fastest): "MyUSB Drive" (and therefore the USB sniffer for speed test 3) is connected directly to the root hub, not to the hub! I have too many hubs...

**Comment by ataradov** [comment]
Voltage regulator should not matter. I just used the one I had in stock. 

I guess I'm missing something with the way those hub commands are sent. It does not really hurt anything, I was just not expecting to see them in the log.

**Comment by marcosribeirobr** [comment]
@wocard2 I'm curious to understand if the problem was solved?

**Comment by wocard2** [comment]
Yes, when connecting the sniffer to my fastest USB port (that xHCI controller is probably provided by the Intel C216 chipset) without hub, high-speed sniffing works just fine.
